src/scalar/dusk.rs

```diff
@@ -262,34 +261,34 @@ impl Scalar {
         res
     }
 
-    /// Creates a `Scalar` from arbitrary bytes by hashing the input with BLAKE2b into a 256-bits
-    /// number, and then converting it into its `Scalar` representation.
-    pub fn from_var_bytes(input: &[u8]) -> Scalar {
+    /// Creates a `Scalar` from arbitrary bytes by hashing the input with
+    /// BLAKE2b into a 512-bits number, and then converting the number into its
+    /// `Scalar` representation by reducing it by the modulo.
+    ///
+    /// This implementation follows the first conversion of
+    /// https://hackmd.io/zV6qe1_oSU-kYU6Tt7pO7Q with concrete numbers:
+    /// ```text
+    /// p = 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001
+    /// p = 52435875175126190479447740508185965837690552500527637822603658699938581184513
+    ///
+    /// s = 2^512
+    /// s = 13407807929942597099574024998205846127479365820592393377723561443721764030073546976801874298166903427690031858186486050853753882811946569946433649006084096
+    ///
+    /// r = 255699135089535202043525422716183576215815630510683217819334674386498370757523
+    ///
+    /// m = 3294906474794265442129797520630710739278575682199800681788903916070560242797
+    /// ```
+    pub fn from_blake2b(input: &[u8]) -> Scalar {
         let state = blake2b_simd::Params::new()
-            .hash_length(32)
+            .hash_length(64)
             .to_state()
             .update(input)
             .finalize();
 
-        let h = state.as_bytes();
-        let mut r = [0u64; 4];
-
-        // will be optmized by the compiler, depending on the available target
-        for i in 0..4 {
-            r[i] = u64::from_le_bytes([
-                h[i * 8],
-                h[i * 8 + 1],
-                h[i * 8 + 2],
-                h[i * 8 + 3],
-                h[i * 8 + 4],
-                h[i * 8 + 5],
-                h[i * 8 + 6],
-                h[i * 8 + 7],
-            ]);
-        }
+        let mut bytes = [0u8; 64];
```

Comment:
It's a waste of cycles to allocate a new array only to later use it as reference to generate `8xu64`. Why not ready the little-endian bytes directly from the generated output of the hash function?

Reply:
I'd love to not need to generate another array. Do you have a suggestion? Mind that `Scalar::from_bytes_wide` expects a reference to a 64 bytes array.

Reply:
Sure!

```rust
let state = blake2b_simd::Params::new()
    .hash_length(64)
    .to_state()
    .update(input)
    .finalize();

let bytes = state.as_bytes();

Scalar::from_u512([
    u64::from_le_bytes(<[u8; 8]>::try_from(&bytes[0..8]).unwrap()),
    u64::from_le_bytes(<[u8; 8]>::try_from(&bytes[8..16]).unwrap()),
    u64::from_le_bytes(<[u8; 8]>::try_from(&bytes[16..24]).unwrap()),
    u64::from_le_bytes(<[u8; 8]>::try_from(&bytes[24..32]).unwrap()),
    u64::from_le_bytes(<[u8; 8]>::try_from(&bytes[32..40]).unwrap()),
    u64::from_le_bytes(<[u8; 8]>::try_from(&bytes[40..48]).unwrap()),
    u64::from_le_bytes(<[u8; 8]>::try_from(&bytes[48..56]).unwrap()),
    u64::from_le_bytes(<[u8; 8]>::try_from(&bytes[56..64]).unwrap()),
])
```

Reply:
As rule of thumb, code duplication should be avoided only when there is no performance impact and is only a matter of code organization